R/visit.R

```diff
@@ -43,3 +43,166 @@ visit_one <- function(pd_flat, funs) {
+#' @return An updated parse table.
+#' @seealso context_to_terminals
+context_towards_terminals <- function(pd_nested,
+                                      passed_lag_newlines,
+                                      passed_indent,
+                                      passed_spaces) {
+  pd_nested$indent <- pd_nested$indent + passed_indent
+  pd_nested$lag_newlines[1] <- pd_nested$lag_newlines[1] + passed_lag_newlines
+  pd_nested$spaces[nrow(pd_nested)] <-
+    pd_nested$spaces[nrow(pd_nested)] + passed_spaces
+  pd_nested
+}
+
+#' Extract terminal tokens
+#'
+#' Turns a nested parse table into a flat parse table. In particular it extracts
+#'   terminal tokens and the following attributes:
+#'
+#'  * lag_newlines
+#'  * indent
+#'  * token
+#'  * text
+#'  * spaces
+#'  * id
+#'  * parent
+#'  * line1
+#' @inheritParams extract_terminals_helper
+#' @importFrom readr type_convert col_integer cols
+extract_terminals <- function(pd_nested) {
+  flat_vec <- extract_terminals_helper(pd_nested) %>%
+    unlist()
+  nms <- list(
+    NULL,
+    c("lag_newlines", "indent", "token", "text", "spaces", "id", "parent", "line1")
+  )
+  flat_tbl <- matrix(flat_vec, ncol = length(nms[[2]]), byrow = TRUE, dimnames = nms) %>%
+    as_tibble() %>%
+    type_convert(
+      col_types = cols(
+        lag_newlines = col_integer(),
+        indent       = col_integer(),
+        spaces       = col_integer()
+      )
+    )
+}
+
+#' Helper to extract terminals
+#'
+#' @param pd_nested A nested parse table.
+extract_terminals_helper <- function(pd_nested) {
+  if (is.null(pd_nested)) return(pd)
+  pmap(list(pd_nested$terminal, pd_nested$token, pd_nested$text,
+            pd_nested$lag_newlines, pd_nested$spaces, pd_nested$indent,
+            pd_nested$id, pd_nested$parent, pd_nested$line1, pd_nested$child),
+       function(terminal, token, text, lag_newlines, spaces, indent, id,
+                parent, line1, child) {
+         if (terminal) {
+           c(lag_newlines, indent, token, text, spaces, id, parent, line1)
+         } else {
+           extract_terminals_helper(child)
+         }
+       })
+}
+
+#' Enrich flattened parse table
+#'
+#' Enriches a flattened parse table with terminals only. In particular, it is
+#'   possible to compute the exact position a token will have (line and column)
+#'   when it will be serialized.
+#' @details Since we have only terminal tokens now, the line on which a token
+#'  starts we also be the line on which it ends. We call `line1` the line on
+#'  which the token starts. `line1` has the same meaning as `line1` that can be
+#'  found in a flat parse table (see [tokenize()]), just that the `line1`
+#'  created by `enrich_terminals()` is the updated version of the former
+#'  `line1`. The same applies for `col1` and `col2`.
+#' @inheritParams choose_indention
+enrich_terminals <- function(flattened_pd, use_raw_indention = FALSE) {
+  flattened_pd$lag_spaces <- lag(flattened_pd$spaces, default = 0)
```

Comment:
uups, changed it elsewhere. Now I think all `default  = 0` are changed to `default = 0L`.